### PR TITLE
Keycloak: add clientscope management

### PR DESCRIPTION
**Pull request by Gaetan2907**
##### SUMMARY
This PR add a module to manage keycloak clientscope, like keycloak_client manage clients

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
Module name is keycloak_clientscope

### CODE REVIEW

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_clientscope
+
+short_description: Allows administration of Keycloak client_scopes via Keycloak API
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:13]:
```suggestion
short_description: Allows administration of Keycloak client_scopes via Keycloak API

version_added: 3.4.0
```

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+    description:
+        type: str
+        description:
+            - Description for this client_scope.
+            - This parameter is not required for updating or deleting a client_scope.
+
+    protocol:
+        description:
+            - Type of client (either C(openid-connect), C(saml) or C(wsfed)).
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:72]: No need to list the choices twice.
```suggestion
            - Type of client.
```

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+            - a list of dicts defining protocol mappers for this client.
+              This is 'protocolMappers' in the Keycloak REST API.
+        aliases:
+            - protocolMappers
+        type: list
+        elements: dict
+        suboptions:
+            protocol:
+                description:
+                    - This is either C(openid-connect), C(saml) or C(wsfed), this specifies for which protocol this protocol mapper
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:87]:
```suggestion
                    - This specifies for which protocol this protocol mapper
```

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+                      is active.
+                choices: ['openid-connect', 'saml', 'wsfed']
+                type: str
+
+            protocolMapper:
+                description:
+                    - The Keycloak-internal name of the type of this protocol-mapper. While an exhaustive list is
+                      impossible to provide since this may be extended through SPIs by the user of Keycloak,
+                      by default Keycloak as of 3.4 ships with at least
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:96]:
```suggestion
                      by default Keycloak as of 3.4 ships with at least:
```

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+                    - Usually a UUID specifying the internal ID of this protocol mapper instance.
+                type: str
+
+            config:
+                description:
+                    - Dict specifying the configuration options for the protocol mapper; the
+                      contents differ depending on the value of I(protocolMapper) and are not documented
+                      other than by the source of the mappers and its parent class(es). An example is given
+                      below. It is easiest to obtain valid config values by dumping an already-existing
+                      protocol mapper configuration through check-mode in the I(existing) field.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:140]: You mean the return value, right?
```suggestion
                      protocol mapper configuration through check-mode in the C(existing) return value.
```

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+                type: dict
+
+    attributes:
+        type: dict
+        description:
+            - A dict of key/value pairs to set as custom attributes for the client_scope.
+            - Values may be single values (e.g. a string) or a list of strings.
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:147]:
```suggestion
            - Values may be single values (for example a string) or a list of strings.
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added a few small comments. Please note that you need tests for this new module as well.

File: plugins/modules/identity/keycloak/keycloak_clientscope.py (hunk 5 above)
Comment (Gaetan2907) [reply plugins/modules/identity/keycloak/keycloak_clientscope.py:142]: Yes indeed, thanks

**Comment by Gaetan2907** [comment]
Thanks for the review ! 
I will add the tests today

File: plugins/modules/identity/keycloak/keycloak_clientscope.py (hunk 4 above)
Comment (Gaetan2907) [reply plugins/modules/identity/keycloak/keycloak_clientscope.py:98]: This change made pep8 through an error so I have removed it. 
Not sure if there is another way of fixing it.

**Comment by felixfontein** [comment]
@kris2kris @laurpaum you might also be interested in this PR.

File: plugins/modules/identity/keycloak/keycloak_clientscope.py [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+            protocol:
+                description:
+                    - This specifies for which protocol this protocol mapper
+                    - is active.
+                choices: ['openid-connect', 'saml', 'wsfed']
+                type: str
+
+            protocolMapper:
+                description:
+                    - The Keycloak-internal name of the type of this protocol-mapper. While an exhaustive list is
```
Comment (felixfontein) [inline plugins/modules/identity/keycloak/keycloak_clientscope.py:96]:
```suggestion
                    - "The Keycloak-internal name of the type of this protocol-mapper. While an exhaustive list is
```

File: plugins/modules/identity/keycloak/keycloak_clientscope.py (hunk 4 above)
Comment (felixfontein) [reply plugins/modules/identity/keycloak/keycloak_clientscope.py:98]:
```suggestion
                      by default Keycloak as of 3.4 ships with at least:"
```

**Review by felixfontein — APPROVED** [review]
I don't use keycloak, but the module code looks ok to me. As nobody seems to want to properly review this, I guess we'll merge it and see if someone complains that it doesn't work as expected :)

**Comment by felixfontein** [comment]
@Gaetan2907 thanks for your contribution!

**Pull request merged**